Assistant: Add prompt rendering mechanism

georgestagg opened the pull request:
- addresses #9563
- addresses https://github.com/posit-dev/positron/issues/10174

This PR has three main aims:

 * Rework how our internal prompts are built, introducing templating and allowing us to move logic from the extension's `.ts` source into `.md` files.
 
 * Create a way for developers to enable and disable parts of Assistant's prompts at runtime, to assist with testing and QA.
 
 * Replace our `llms.txt` implementation and instead integrate Assistant with Code OSS's notion of prompt files, custom instructions, modes, & commands. (See [here](https://code.visualstudio.com/docs/copilot/customization/custom-instructions) for details).

Also included are some general fixes to prompting and tool calls, including the changes in #9991 and #9974. I will rebase those specific changes away from this branch as they hit `main`.

### Internal prompt files

Assistant's prompt files may now have a yaml header with a `mode` property indicating in which modes the prompt should be included. Prompts are concatenated in the order defined by the `order` yaml property.

Templating is done using [Squirrely](https://squirrelly.js.org), and this allows us to move some of the branching logic (e.g. different wording in the "inline editor", or when streaming edits is enabled) into the prompt files. This is the same template engine Databot currently uses, aligning us there.

The default prompt file has been split into different sections, reflecting the purpose of each section.

Prompts for commands have the `command` property, and also a `mode` property, defining the slash command and for which modes the command is active. A command's prompt document(s) is appended to the system prompt for the currently active mode.

## Developer prompt management

A new command has been added, gated by `isDevelopment` (Q: What is this value in daily builds?). When enabled, the command allows a user to turn on and off parts of the Assistant system prompts.

This allows for a quick way to test the effect of individual prompt sections, and even return to the stock model if required:

<img width="630" height="289" alt="Screenshot 2025-10-16 at 11 49 59" src="https://github.com/user-attachments/assets/bcd51845-18fd-457f-8ff9-0a3fdf8f8393" />

It might be nice if this was on for daily builds, but should definitely be off on release builds.

Note: Positron context info is still added even if all prompts are off.

## LLM instructions

Our special handling for `llms.txt` has been removed. With this PR it is added to requests using a different mechanism (next section).

## Custom prompt files

Upstream Code OSS has support for custom instructions, commands and chat modes. This PR makes this all work better with Positron. It also adds our own directories for the custom files, under `.positron`, to separate from the GitHub Copilot Chat naming.

Here are some examples:

### Custom Command
Use the cog icon in Assistant, select Prompt Files, then create a new prompt file in the `.positron/prompts` folder.

<img width="920" height="325" alt="Screenshot 2025-10-16 at 11 56 13" src="https://github.com/user-attachments/assets/e77b3e90-9970-438e-bfab-d0808369c4f4" />

Here is an example defining a custom `/spelling` command, saved as `./positron/prompts/spelling.prompt.md

```
---
mode: agent
---
Check the document for spelling errors.
```

This should then be available as a command in the given chat mode:

<img width="565" height="654" alt="Screenshot 2025-10-16 at 11 59 55" src="https://github.com/user-attachments/assets/c52bdabd-f446-40a1-b06d-d3279c8314a4" />

### Custom Chat Mode

<img width="808" height="237" alt="Screenshot 2025-10-16 at 12 00 38" src="https://github.com/user-attachments/assets/e360bc57-d187-43b6-a507-cbca791d1883" />

Example: `.positron/chatmodes/pirate.chatmode.md`:

```
---
description: 'Speak like a pirate'
tools: []
---
You must speak like a pirate in this mode.
```

<img width="573" height="571" alt="Screenshot 2025-10-16 at 12 02 17" src="https://github.com/user-attachments/assets/444e8cf9-9e3d-4b91-9ff7-bed642bdd3e1" />

### Custom instructions

This already works as expected, in that `AGENTS.md` is loaded automatically into the context if it exists in the root of the project.

This PR tweaks the upstream mechanism to also include any of:

 -  'agents.md'
 - 'agent.md'
 - 'positron.md'
 - 'claude.md'
 - 'gemini.md'
 - 'llms.txt'

### Project tree tool
I have simplified the project tree tool a lot. I found in testing the model often struggles to infer paths from the previous array-based output format. The tool now outputs paths as a newline separated list:

```
foo/bar/baz.txt
foo/bar.txt
abc/def.txt
efg.txt
```

If the number of entries is too long, the list will be limited by removing the longest paths first.

## Still TODO

* With the introduction of a prompt rendering mechanism it would probably be good to have some tests in place specifically for it.

* Chat export can currently be used to get a dump of a full LLM conversation and tool calling loop, but it would be good if we had a testing entry point that allows us to test tools without starting a full blown Positron session. Probably that would be complex enough for a followup PR.



### Release Notes

#### New Features

- Assistant: enable custom chat modes, instructions and prompt files (#9563)
- Assistant: simplify project tree tool output structure (#10174)

georgestagg committed:
Assistant: Refactor command prompt handling

georgestagg committed:
Assistant: Rework prompting for new templating structure

georgestagg committed:
Assistant: Prompt descriptions and quickpick manager

georgestagg committed:
Assistant: Simplify project tree tool

georgestagg committed:
Assistant: Allow empty prompt when all deselected

georgestagg committed:
Assistant: Include .positron prompt files and additional AGENTS.md

georgestagg committed:
Assistant: Only show model management quickpick in dev mode

georgestagg committed:
Remove no longer required llms.txt test

georgestagg committed:
Assistant: Add instruction files in an "instructions" node

georgestagg commented on src/vs/workbench/contrib/chat/common/promptSyntax/computeAutomaticInstructions.ts:
```diff
@@ -202,7 +202,20 @@ export class ComputeAutomaticInstructions {
+					/*
 					const agentMd = root.stat.children.find(c => c.isFile && c.name.toLowerCase() === 'agents.md');
+					*/
+					const agentMd = root.stat.children.find(
```
Should we attach multiple instruction files? How do we handle symlinks?

georgestagg committed:
Update assistant tests for PromptRenderer

georgestagg committed:
Assistant: Apply prompt renderer when working with GitHub provider

georgestagg committed:
Remove use of Positron Context in prompt renderer

georgestagg committed:
Move additional search path to under .vscode

sharon-wang commented on extensions/positron-assistant/src/promptRender.ts:
```diff
@@ -0,0 +1,395 @@
+		{ label: 'Editor', description: 'Inline editor chat' },
+		{ label: 'Terminal', description: 'Inline Terminal chat' },
+		{ label: 'Notebook', description: 'Notebook chat' },
+		{ label: 'Miscelleaneous', kind: vscode.QuickPickItemKind.Separator },
```
spelling nit

```suggestion
		{ label: 'Miscellaneous', kind: vscode.QuickPickItemKind.Separator },
```

sharon-wang reviewed (commented):
some initial thoughts before I test this out early next week!

jonvanausdeln commented on extensions/positron-assistant/src/test/participants.test.ts:
```diff
@@ -305,36 +308,6 @@ ${attachmentsText}
 			});
 	});
 
-	test('should include llms.txt instructions', async () => {
```
> Can we add some basic tests now that we're removing this one? CC @jonvanausdeln for e2e testing thoughts too

I've been thinking about that, and I think I may be able to add some using the echo provider, and just parse out the prompts to compare against expected.

georgestagg committed:
Assistant: remove activation steering

georgestagg committed:
Minor tweak to prompt metadata merging logic

georgestagg committed:
Update extensions/positron-assistant/src/promptRender.ts

Co-authored-by: sharon <sharon-wang@users.noreply.github.com>
Signed-off-by: George Stagg <georgestagg@gmail.com>

georgestagg committed:
Rename shiny participant prompt file

georgestagg committed:
Internationalise prompt management description strings

georgestagg commented:
I've rebased on `main` with the latest upstream and force pushed, so if you've checked this out you might need to remove and re-pull.

I've also added a couple of minor changes. Most importantly removing the activation steering prompt for the reasons described in #10137.

sharon-wang reviewed (approved):
This is looking good to me!

Here's a basic test using a custom chat mode, custom prompt, and 3 custom instructions files at once.

<img width="1683" height="985" alt="Image" src="https://github.com/user-attachments/assets/e0663d55-bfb8-4ac9-aea2-e8bf6acf37e6" />

A note that the project tree tool seems much faster now!

---

> A new command has been added, gated by isDevelopment (Q: What is this value in daily builds?). When enabled, the command allows a user to turn on and off parts of the Assistant system prompts.

If gated by `isDevelopment`, they will only be available in local dev builds, as the daily builds are release builds too. I don't think we currently have a mechanism for flags affecting daily-but-not-monthly release-builds, so I think we'll have to enable for dev mode only, or gate via a setting (but that would make it available in the monthly release builds too)

---

@jonvanausdeln would you be up for adding some tests to this PR or as a follow-up?
